### PR TITLE
`Iris`: Disable chatbots in exam exercises

### PR DESCRIPTION
**Pull request by Hialus**
> [!NOTE]
> Only deploy on TS9

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I documented the TypeScript code using JSDoc style.

### Motivation and Context

Iris was not yet blocked for exam exercises. As you don't want a chatbot that could potentially give out code in the exams, we want to disable this completely.

### Description

This PR adds several checks on the server and client to disable Iris for exams.

### Steps for (Exam Mode) Testing

Prerequisites:
- 1 Instructor

1. Log in to Artemis
2. Go to a programming exercise where Iris is enabled (e.g. https://artemis-test9.artemis.cit.tum.de/courses/6/exercises/23)
3. Iris still works
4. Create a new exam in a course where Iris is enabled and add a programming exercise
5. The programming exercise edit page does not show an Iris button
6. The participation view of a student does not show an Iris button

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2
#### Exam Mode Test
- [x] Test 1
- [ ] Test 2

### Test Coverage

Unchanged.

### CODE REVIEW

**Review by Strohgelaender — DISMISSED** [review]
code

**Review by DominikRemo — DISMISSED** [review]
Code LGTM

**Review by MichaelOwenDyer — APPROVED** [review]

**Review by RY997 — DISMISSED** [review]
When deploying to ts9, I got this. I think maybe it's because of the docker issue that Stephan mentioned. 
The code LTGM.
Btw, the deployment-error label didn't replace the lock label this time, is this intentional?
image

File: src/main/java/de/tum/in/www1/artemis/service/iris/IrisSessionService.java [hunk 1]
```diff
@@ -58,6 +58,9 @@ public void checkIsIrisActivated(IrisSession session) {
      * @return The created session
      */
     public IrisSession createChatSessionForProgrammingExercise(ProgrammingExercise exercise, User user) {
+        if (exercise.isExamExercise()) {
+            throw new BadRequestException("Iris is not supported for exam exercises");
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/iris/IrisSessionService.java:62]: We (@bassner and I) had something similar in one of my PRs and decided this should be a 409 (Conflict) as this is not a client error.

File: src/main/java/de/tum/in/www1/artemis/service/iris/IrisSettingsService.java [hunk 2]
```diff
@@ -360,6 +361,9 @@ public IrisSettings saveIrisSettings(Course course, IrisSettings settings) {
      * @return the saved Iris settings
      */
     public IrisSettings saveIrisSettings(ProgrammingExercise exercise, IrisSettings settings) {
+        if (exercise.isExamExercise()) {
+            throw new BadRequestException("Iris is not supported for exam exercises");
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/iris/IrisSettingsService.java:365]: same here

File: src/main/java/de/tum/in/www1/artemis/service/iris/session/IrisChatSessionService.java [hunk 3]
```diff
@@ -123,6 +123,9 @@ public void requestAndHandleResponse(IrisSession session) {
         if (!(fullSession instanceof IrisChatSession chatSession)) {
             throw new BadRequestException("Trying to get Iris response for session " + session.getId() + " without exercise");
         }
+        if (((IrisChatSession) fullSession).getExercise().isExamExercise()) {
+            throw new BadRequestException("Iris is not supported for exam exercises");
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/iris/session/IrisChatSessionService.java:127]: same here

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/iris/IrisSessionService.java (hunk 1 above)
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/service/iris/IrisSessionService.java]: Changed it 👍🏻

**Review by DominikRemo — APPROVED** [review]

**Review by MaximilianAnzinger — APPROVED** [review]
Code LGTM, thx for incorporating my suggestions

**Pull request merged**